File: python/app/v1/items.py
from abc import ABC, abstractmethod
from dataclasses import dataclass

@dataclass(frozen=True, slots=True)
class Id:
    value: str

class Item(ABC):
    def __init__(self, id: Id, title: str) -> None:
        self._id, self._title = id, title

    @property
    def id(self) -> Id: 
        return self._id

    @property
    def title(self) -> str: 
        return self._title

    @abstractmethod
    def days_allowed(self) -> int: ...

    @abstractmethod 
    def description(self) -> str: ...

class Book(Item):
    def days_allowed(self) -> int: 
        return 14
    def description(self) -> str:
        return f"{self.title} ({self.days_allowed()} days)"

class Dvd(Item):
    def days_allowed(self) -> int: 
        return 7
    def description(self) -> str:
        return f"{self.title} ({self.days_allowed()} days)"

File: rust/src/v1/items.rs
pub trait Item {
    fn id(&self) -> &str;

    fn title(&self) -> &str;

    fn days_allowed(&self) -> u32;
    #[allow(dead_code)]
    fn description(&self) -> String {
        format!("{} ({} days)", self.title(), self.days_allowed())
    }
}

pub struct Book {
    id: String,
    title: String,
}

impl Book {
    pub fn new(id: &str, title: &str) -> Self {
        Book {
            id: id.to_string(),
            title: title.to_string(),
        }
    }
}

impl Item for Book {
    fn id(&self) -> &str {
        &self.id
    }

    fn title(&self) -> &str {
        &self.title
    }

    fn days_allowed(&self) -> u32 {
        14
    }

    fn description(&self) -> String {
        format!("{} ({} days)", self.title, self.days_allowed())
    }
}

pub struct Dvd {
    id: String,
    title: String,
}

impl Dvd {
    pub fn new(id: &str, title: &str) -> Self {
        Dvd {
            id: id.to_string(),
            title: title.to_string(),
        }
    }
}

impl Item for Dvd {
    fn id(&self) -> &str {
        &self.id
    }

    fn title(&self) -> &str {
        &self.title
    }

    fn days_allowed(&self) -> u32 {
        7
    }

    fn description(&self) -> String {
        format!("{} ({} days)", self.title, self.days_allowed())
    }
}
